### DISCUSSION TITLE
CTR (context-aware) but also sequential DIEN model usage.

### DISCUSSION THREAD

**Discussion by Jeriousman** [RUCAIBox/RecBole#1212] [Q&A] [answered]
CTR (context-aware) and sequential model should have different default settings in the documentation but DIEN is one with both CTR (context-aware) and sequential one applied in it. So my start point was your default settings for DIEN in

```
from recbole.quick_start import run_recbole
run_recbole(model='DIEN', dataset='ml-100k',)
```
and I check the default config values to benchmark them.  Below is the default config

```
General Hyper Parameters:
gpu_id = 0
use_gpu = True
seed = 2020
state = INFO
reproducibility = True
data_path = /home/hojun/Python/Models/Recommender_systems/RecBole/recbole/config/../dataset_example/ml-100k
checkpoint_dir = saved
show_progress = True
save_dataset = True
dataset_save_path = None
save_dataloaders = True
dataloaders_save_path = None
log_wandb = False

Training Hyper Parameters:
epochs = 1
train_batch_size = 2048
learner = adam
learning_rate = 0.001
neg_sampling = {'uniform': 1}
eval_step = 1
stopping_step = 10
clip_grad_norm = None
weight_decay = 0.0
loss_decimal_place = 4

Evaluation Hyper Parameters:
eval_args = {'split': {'LS': 'valid_and_test'}, 'order': 'TO', 'mode': 'uni100', 'group_by': 'user'}
repeatable = True
metrics = ['AUC', 'LogLoss']
topk = [10]
valid_metric = AUC
valid_metric_bigger = True
eval_batch_size = 4096
metric_decimal_place = 4

Dataset Hyper Parameters:
field_separator = 	
seq_separator =  
USER_ID_FIELD = user_id
ITEM_ID_FIELD = item_id
RATING_FIELD = rating
TIME_FIELD = timestamp
seq_len = None
LABEL_FIELD = label
threshold = None
NEG_PREFIX = neg_
load_col = {'inter': ['user_id', 'item_id', 'rating', 'timestamp'], 'user': ['user_id', 'age', 'gender', 'occupation'], 'item': ['item_id', 'release_year']}
unload_col = None
unused_col = None
additional_feat_suffix = None
rm_dup_inter = None
val_interval = None
filter_inter_by_user_or_item = True
user_inter_num_interval = None
item_inter_num_interval = None
alias_of_user_id = None
alias_of_item_id = None
alias_of_entity_id = None
alias_of_relation_id = None
preload_weight = None
normalize_field = None
normalize_all = None
ITEM_LIST_LENGTH_FIELD = item_length
LIST_SUFFIX = _list
MAX_ITEM_LIST_LENGTH = 50
POSITION_FIELD = position_id
HEAD_ENTITY_ID_FIELD = head_id
TAIL_ENTITY_ID_FIELD = tail_id
RELATION_ID_FIELD = relation_id
ENTITY_ID_FIELD = entity_id
benchmark_filename = None

Other Hyper Parameters: 
wandb_project = recbole
require_pow = False
embedding_size = 10
mlp_hidden_size = [256, 256, 256]
dropout_prob = 0
pooling_mode = mean
gru_type = AUGRU
alpha = 1
MODEL_TYPE = ModelType.SEQUENTIAL
MODEL_INPUT_TYPE = InputType.POINTWISE
eval_type = EvaluatorType.VALUE
device = cuda
train_neg_sample_args = {'strategy': 'by', 'by': 1, 'distribution': 'uniform', 'dynamic': 'none'}
eval_neg_sample_args = {'strategy': 'by', 'by': 100, 'distribution': 'uniform'}
```

Here my first question arises.

1. LABEL_FIELD is set to label but ml-100k dataset does not have label but rating field. The label field name is mismatched but the code runs without error. How is the mechanism of this working when name 'label' doesnt exist in the dataset? 

2. For CTR purpose, I do not want to use 'rating' field at all and see every datapoint as positive one. And then do negative sampling. then now I do not have anything for ```LABEL_FIELD```, and no 'rating' feature used in ```load_col = {'inter': ['user_id', 'item_id', 'rating', 'timestamp']```. How should I set them?

3. some feature columns like gender, sex and item description just for example dont have to be sequential but item and category should be sequential data. How can I set the specific feature (item and category) to be treated and formed as sequential data while gender, sex and item description are treated non-sequential but general data in config?

Thank you in advance.

**Comment by Ethan-TZ** [comment, marked as answer]
@Jeriousman Hello, thanks for your attention to RecBole!

1. For movie-lens dataset, we label the data by setting thresholds. i.e., we set 3 as the threshold of rating by default, and the interactions will be labeled as positive if their rating no less than 3. You can refer to [LABEL OF DATA](https://recbole.io/docs/user_guide/data/label_of_data.html) for more details.
2. We don't support negative sampling for context-aware models, because it is hard to construct context information(e.g. click-time) for the negative samples.
3. For feature fields in **.inter** file except user_id,we will do data augmentation to construct the sequential features according to user history interaction. This process is automatic, so you only need to put the features which are treated and formed as sequential data into the **.inter** file. You can refer to [sequential dataset](https://recbole.io/docs/recbole/recbole.data.dataset.sequential_dataset.html?highlight=augmentation) for more help. If feature field is already sequential(e.g. class in ml-100k), please indicate the feature type end with '_seq' (e.g. class:token_seq).

**Reply by Jeriousman** [reply]
Thank you very much for your feedback.
